File: tree/95.unique-binary-search-trees-ii.go
/*
 * @lc app=leetcode id=95 lang=golang
 *
 * [95] Unique Binary Search Trees II
 */

// @lc code=start
// *
// Definition for a binary tree node.
// type TreeNode struct {
// 	Val   int
// 	Left  *TreeNode
// 	Right *TreeNode
// }

func generateTrees(n int) []*TreeNode {
	if n == 0 {
		return []*TreeNode{}
	}
	return helper(1, n)
}

func helper(l, r int) []*TreeNode {
	result := []*TreeNode{}
	if l > r {
		result = append(result, nil)
		return result
	}
	for i := l; i <= r; i++ {
		for _, left := range helper(l, i-1) {
			for _, right := range helper(i+1, r) {
				root := TreeNode{Val: i}
				root.Left = left
				root.Right = right
				result = append(result, &root)
			}
		}
	}
	return result
}

// @lc code=end

File: tree/95.unique-binary-search-trees-ii.py
from typing import List

#
# @lc app=leetcode id=95 lang=python3
#
# [95] Unique Binary Search Trees II
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def generateTrees(self, n: int) -> List[TreeNode]:
        if n == 0:
            return []

        return self.helper(1, n)

    def helper(self, l: int, r: int) -> List[TreeNode]:
        result: List[TreeNode] = []
        if l > r:
            result.append(None)
            return result
        for i in range(l, r + 1):
            for left in self.helper(l, i - 1):
                for right in self.helper(i + 1, r):
                    root = TreeNode(val=i, left=left, right=right)
                    result.append(root)
        return result


# @lc code=end
